apps/ejabberd/src/ejabberd_c2s.erl

```diff
@@ -1161,10 +1155,65 @@ handle_info({force_update_presence, LUser}, StateName,
             StateData
     end,
     {next_state, StateName, NewStateData};
-handle_info({send_filtered, Feature, From, To, Packet}, StateName, StateData) ->
+handle_info(resume_timeout, resume_session, StateData) ->
+    {stop, normal, StateData};
+handle_info(check_buffer_full, StateName, StateData) ->
+    case is_buffer_full(StateData#state.stream_mgmt_buffer_size,
+                        StateData#state.stream_mgmt_buffer_max) of
+        true ->
+            Err = ?RESOURCE_CONSTRAINT_ERR((StateData#state.lang),
+                                           <<"too many unacked stanzas">>),
+            send_element(StateData, Err),
+            send_trailer(StateData),
+            {stop, normal, StateData};
+        false ->
+            fsm_next_state(StateName,
+                           StateData#state{stream_mgmt_constraint_check_tref = undefined})
+    end;
+handle_info({store_session_info, User, Server, Resource, KV, _FromPid}, StateName, StateData) ->
+    ejabberd_sm:store_info(User, Server, Resource, KV),
+    fsm_next_state(StateName, StateData);
+handle_info(Info, StateName, StateData) ->
+    handle_incoming_message(Info, StateName, StateData).
+
+maybe_terminate(Acc) ->
+    % instead of this, we will pass on the accumulator, only replacing
+    % 'element' with 'to_send', if present
+    case mongoose_acc:is_acc(Acc) of
+        true ->
+            mongoose_acc:terminate(Acc, received, ?FILE, ?LINE);
+        false ->
+            ?ERROR_MSG("Hey, it should be accumulator here! ~p", [Acc]),
+            Acc
+    end.
+
+%%% incoming messages
+handle_incoming_message({send_text, Text}, StateName, StateData) ->
+    ?ERROR_MSG("{c2s:send_text, Text}: ~p~n", [{send_text, Text}]), % is it ever called?
+    send_text(StateData, Text),
+    ejabberd_hooks:run(c2s_loop_debug, [Text]),
+    fsm_next_state(StateName, StateData);
+handle_incoming_message({broadcast, Acc}, StateName, StateData) ->
+    Broadcast = maybe_terminate(Acc),
+    ejabberd_hooks:run(c2s_loop_debug, [{broadcast, Broadcast}]),
+    ?DEBUG("broadcast=~p", [Broadcast]),
+    Res = handle_routed_broadcast(Broadcast, StateData),
+    handle_broadcast_result(Res, StateName, StateData);
+handle_incoming_message({route, From, To, Acc0}, StateName, StateData) ->
+    % since we are now in the other user's session some cached data are not valid
+    % anymore (e.g. privacy_check), they have to be removed somewher between
+    % sender and recipient. One might argue, possibly rightly, that it should be
+    % stripped before sending; the reason I do it here is that when we send
+    % an acc out of c2s it returns and we might still use the cached data.
```

Comment:
No, not at the moment. I just have a gut feeling it might be useful. There are more cases along these lines, which need some thinking - for example, a stanza which came in is 'element', and it is sent out, unless we want to send something else, then there is the 'to_send' key. But after routing it out I have to remove 'to_send', because later on I might want to send out the original stanza (it is a real case). I might be worthwhile, however, to store the 'to_send' somewhere else so that eventually we can trace what the accumulator processing has sent out.